libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/internal/impl/MapboxAudioGuidanceServices.kt

```diff
@@ -47,4 +64,16 @@ class MapboxAudioGuidanceServices {
     fun configOwner(context: Context): NavigationConfigOwner = NavigationConfigOwner(context)
 
     fun dataStoreOwner(context: Context) = NavigationDataStoreOwner(context)
+
+    fun registerVoiceInstructionsPlayerObserver(observer: VoiceInstructionsPlayerObserver) {
+        voiceInstructionsPlayerObservers.add(observer)
```

Comment:
A problem I am seeing here is that in the event a `voiceInstructionPlayer` is already created and the developer later invokes this `registerVoiceInstructionsPlayerObserver` call, they won't receive a callback with the existing `voiceInstructionPlayer` instance. Hence, they would always have to register this observer inside onCreate to make sure that they do receive the instance when it is created.

Reply:
I am thinking if we can somehow create this `player` outside of this `combine flow`. It makes sense why it lives in a flow because one needs access to `language` to create the `voiceInstructionPlayer`. This is not available until a route is requested using `mapboxNavigation`.

I see this as a limitation in `VoiceInstructionPlayer` offered from the SDK itself. We should be able to create a `voiceInstructionPlayer` using any language we want and change the language at runtime with/without creating a new instance. I am thinking we should take a stab at it. This way in Drop-In UI, we can do the following:

- Always create a `voiceInstructionPlayer` using the language coming from `configOwner!!.language()` or `locale` when the application starts.
- Upon availability of a route, use `mapboxNavigation` to obtain the new language coming from `directionsRoute` and either update the `voiceInstructionPlayer` instance created above to use this language instead or create a new instance.

Wdyt?
@Guardiola31337 @kmadsen @LukasPaczos

Reply:
> A problem I am seeing here is that in the event a voiceInstructionPlayer is already created and the developer later invokes this registerVoiceInstructionsPlayerObserver call, they won't receive a callback with the existing voiceInstructionPlayer instance

it can be fixed with if we save a link to a current player and pass it to each new observer

> We should be able to create a voiceInstructionPlayer using any language we want and change the language at runtime with/without creating a new instance. 

it makes sense, btw at the moment when I go FD -> AG -> FD I get a new player on each step, because language changes (`ru` / `Ru-ru`)

Reply:
>it makes sense, btw at the moment when I go FD -> AG -> FD I get a new player on each step, because language changes (ru / Ru-ru)

Do you mean language changes b/w FD and AG state? I am assuming you are changing the language manually b/w these state? Because default Drop-In UI doesn't do so

Reply:
@abhishek1508 yes, language changes between AG / FD, I do not change it manually. So, player is recreated on each AG/FD switch (`configOwner!!.language()` is different from `directionsRoute  language`: `ru / Ru-ru`)

Reply:
> it makes sense, btw at the moment when I go FD -> AG -> FD I get a new player on each step, because language changes (ru / Ru-ru)

That's exactly what I was talking about on Slack.

> Always create a voiceInstructionPlayer using the language coming from configOwner!!.language() or locale when the application starts.
Upon availability of a route, use mapboxNavigation to obtain the new language coming from directionsRoute and either update the voiceInstructionPlayer instance created above to use this language instead or create a new instance.

It looks reasonable. I don't see any problem with creating a new instance when language changes but I find it weird that language changes between FD and AG. With the current API it can change but it's usually some locale clarification, not a drastic change. I think in most cases we should not change instances when switching from FD to AG and back.
Moreover, then not using Drop-In we create a val instructionPlayer in all of our example activities. So it does not change at all.

Another question is do we need voiceInstructionsPlayer in any state other than AG. But if it's just one instance of a player across the whole lifecycle, there's no need to complicate things.

Reply:
> @ abhishek1508 yes, language changes between AG / FD, I do not change it manually. So, player is recreated on each AG/FD switch (configOwner!!.language() is different from directionsRoute language: ru / Ru-ru)

Could we make it so the language stays the same through AG/FD? maybe this could be done in the `configOwner`? because those are the same languages 🤔

Reply:
Also, would anyone notice? What's the actual issue?

Reply:
> What's the actual issue?

It looks weird that we have to reinitialize everything when the language did not actually change. Wasting resources and stuff.
Also, I found an issue this situation causes (about first instruction not being played in offline mode). See Slack thread for more details.

Reply:
> It looks weird that we have to reinitialize everything when the language did not actually change. Wasting resources and stuff.

There are many places to optimize resources and this may not be a very impactful optimization. Optimize for making it work first.